xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/page/events/PageNotificationEventListener.java

```diff
@@ -0,0 +1,155 @@
+
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+
+/**
+ * Default implementation of {@link PageNotificationEventDescriptorManager}.
+ *
+ * @version $Id$
+ * @since 9.5RC1
+ */
+@Component
+@Singleton
+@Named(PageNotificationEventListener.NAME)
+public class PageNotificationEventListener extends AbstractEventListener
+{
+    /**
+     * The listener name.
+     */
+    public static final String NAME = "Page Notification Event Listener";
+
+    @Inject
+    private ObservationManager observationManager;
+
+    @Inject
+    private PageNotificationEventDescriptorManager pageNotificationEventDescriptorManager;
+
+    @Inject
+    private TemplateManager templateManager;
+
+    /**
+     * Constructs a {@link PageNotificationEventListener}.
+     */
+    public PageNotificationEventListener()
+    {
+        super(NAME, AllEvent.ALLEVENT);
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        // Filter the event descriptors concerned by the event, then create the concerned events
+        for (PageNotificationEventDescriptor descriptor : pageNotificationEventDescriptorManager.getDescriptorList())
+        {
+            // If the event is expected by our descriptor
+            if (descriptor.getEventTriggers().contains(event.getClass().getCanonicalName())
+                    && checkXObjectCondition(descriptor, source)
+                    && this.evaluateVelocityTemplate(descriptor.getAuthorReference(),
+                    descriptor.getValidationExpression())) {
+                observationManager.notify(
+                        new PageNotificationEvent(descriptor),
+                        "org.xwiki.platform:xwiki-platform-notifications-api", source);
+            }
+        }
+    }
+
+    /**
+     * Ensure that the given the source matches what the descriptor needs.
+     * If the source is an instance of XWikiDocument, will check if the document contains the XObject specified in
+     * the descriptor.
+     *
+     * @param descriptor the event descriptor
+     * @param source the event source
+     * @return
+     */
+    private boolean checkXObjectCondition(PageNotificationEventDescriptor descriptor, Object source) {
+        if (StringUtils.isBlank(descriptor.getObjectType())) {
+            // If no XObject is specified in the Object Type field, we don’t need to check the source.
+            return true;
+        } else if (source instanceof XWikiDocument) {
+            XWikiDocument document = (XWikiDocument) source;
+            Map<DocumentReference, List<BaseObject>> documentXObjects = document.getXObjects();
+            /*  We can’t create a DocumentReference when only using descriptor.objectType, so we will have to
+                    iterate through the map */
+            for (DocumentReference documentReference : documentXObjects.keySet())  {
+                if (documentReference.toString().equals(descriptor.getObjectType())) {
```

Comment:
It's not our best practice to use toString() directly, because the implementation can change.
Instead, you should resolve `getObjectType` to a document reference and compare the references.

You should actually use a Local Serializer, don't you think @tmortagne?